File: java/src/org/revenj/patterns/ServiceLocator.java
package org.revenj.patterns;

import java.lang.reflect.Type;
import java.util.NoSuchElementException;
import java.util.Optional;

public interface ServiceLocator {
	Optional<Object> tryResolve(Type type);

	default <T> T resolve(Class<T> manifest) {
		Optional<Object> found = tryResolve(manifest);
		if (!found.isPresent()) {
			throw new NoSuchElementException(manifest.getName());
		}
		return (T) found.get();
	}

	default <T> T resolve(GenericType<T> genericType) {
		Optional<Object> found = tryResolve(genericType.type);
		if (!found.isPresent()) {
			throw new NoSuchElementException(genericType.type.getTypeName());
		}
		return (T) found.get();
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/patterns/ServiceLocator.scala
package net.revenj.patterns

import java.lang.reflect.Type

import net.revenj.Utils

import scala.reflect.runtime.universe.TypeTag
import scala.util.Try

/** Service for resolving other services.
  * One locator per project should be used.
  *
  * When multiple projects are used, locator must be passed around
  * to resolve appropriate service.
  *
  * Custom classes can be resolved if their dependencies can be satisfied.
  */
trait ServiceLocator {
  /** Resolve a service registered in the locator.
    *
    * @tparam T Type info
    * @return registered implementation
    */
  def resolve[T: TypeTag]: T = {
    val result = tryResolve[T]
    result.getOrElse(throw result.failed.get)
  }

  def tryResolve[T: TypeTag]: Try[T]

  private[revenj] def resolve(tpe: Type): Try[AnyRef]
  private[revenj] def resolve[T](container: Class[T], argument: Type, arguments: Type*): Try[T] = {
    resolve(Utils.makeGenericType(container, argument, arguments: _*)).map(_.asInstanceOf[T])
  }
}
